ibm/resource_ibm_securitygroup_rule.go

```diff
@@ -0,0 +1,196 @@
+		if strings.EqualFold(x, value) {
+			return true, x
+		}
+	}
+	return false, value
+}
+
+func findMatchingRule(d *schema.ResourceData,
+	service services.Network_SecurityGroup) (*datatypes.Network_SecurityGroup_Rule, error) {
+
+	valid, direction := formatString(d.Get("direction").(string), ALLOWED_DIRECTIONS)
+	if !valid {
+		return nil, fmt.Errorf("Invalid direction. %s", direction)
+	}
+	d.Set("direction", direction)
+
+	valid, ethertype := formatString(d.Get("ethertype").(string), ALLOWED_ETHERTYPES)
+	if !valid {
+		return nil, fmt.Errorf("Invalid ethertype. %s", ethertype)
+	}
+	d.Set("ethertype", ethertype)
+
+	portrangemin := d.Get("portrangemin").(int)
+	portrangemax := d.Get("portrangemax").(int)
+	if portrangemin < portrangemax {
+		return nil, fmt.Errorf("portrange max should be greater or equal to min")
+	}
+
+	valid, protocol := formatString(d.Get("protocol").(string), ALLOWED_PROTOCOLS)
+	if !valid {
+		return nil, fmt.Errorf("Invalid protocol. %s", protocol)
+	}
+	d.Set("protocol", protocol)
+
+	securitygroupid := d.Get("securitygroupid").(int)
+	log.Printf("[INFO] %s %d %d %s %d", ethertype, portrangemin, portrangemax, protocol, securitygroupid)
+
+	rules, err := service.Id(securitygroupid).GetRules()
+	if err != nil {
+		return nil, fmt.Errorf("Error fetching information for Security Group Rule: %s", err)
+	}
+	log.Printf("[INFO] rules %s", rules)
+
+	for _, rule := range rules {
+		if direction == *rule.Direction &&
+			ethertype == *rule.Ethertype &&
+			portrangemin == *rule.PortRangeMin &&
+			portrangemax == *rule.PortRangeMax &&
+			protocol == *rule.Protocol {
+			return &rule, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func resourceIBMSecurityGroupRuleCreate(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	service := services.GetNetworkSecurityGroupService(sess)
+
+	matchingrule, err := findMatchingRule(d, service)
+	if err != nil {
+		return err
+	}
+
+	if matchingrule != nil {
+		log.Printf("[INFO] rule exists")
+		d.SetId(fmt.Sprintf("%d", *matchingrule.Id))
+		return nil
+	}
+
+	opts := []datatypes.Network_SecurityGroup_Rule{
+		{
+			Direction:    sl.String(d.Get("direction").(string)),
+			Ethertype:    sl.String(d.Get("ethertype").(string)),
+			PortRangeMin: sl.Int(d.Get("portrangemin").(int)),
+			PortRangeMax: sl.Int(d.Get("portrangemax").(int)),
+			Protocol:     sl.String(d.Get("protocol").(string)),
+		},
+	}
+	log.Println("[INFO] creating security group rule")
+	securitygroupid := d.Get("securitygroupid").(int)
+	_, err = service.Id(securitygroupid).AddRules(opts)
+	if err != nil {
+		return fmt.Errorf("Error creating Security Group Rule: %s", err)
+	}
+
+	return resourceIBMSecurityGroupRuleRead(d, meta)
+}
+
+func resourceIBMSecurityGroupRuleRead(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	service := services.GetNetworkSecurityGroupService(sess)
+
+	matchingrule, err := findMatchingRule(d, service)
```

Comment:
We don't have API to read back a rule based on its id? If so we should use that.
If not then, what happens if the id matches but other fields are different. In that case we shouldn't throw error that rule doesn't exist but we should read back all other attributes of that rule so that `terraform` can correctly show what has changed if at all in the form of a diff on next plan or so. Possibly since the fields are not updatable this might not happen. We need to rethink this.

Reply:
Since we already have the rule id here can't we leverage that?

```go

	rules, err := service.
		Mask("id,portRangeMax").
		Filter(
			filter.Build(
				filter.Path("rules.id").Eq(d.Id()),
			),
		).Id(id).GetRules()
```
This will give all the info which we should setback.